augur/filters/subsample.py

```diff
@@ -0,0 +1,165 @@
+from collections import defaultdict
+import functools
+import random
+
+from augur.filters.base_filter import BaseFilter
+
+
+def is_enabled(args):
+    # TODO raise if one specified without the other
+    return args.group_by or args.sequences_per_group
+
+
+def add_arguments(parser):
+    parser.add_argument(
+        "--sequences-per-group",
+        type=int,
+        help="subsample to no more than this number of sequences per category",
+    )
+    parser.add_argument(
+        "--group-by",
+        nargs="+",
+        help="categories with respect to subsample; two virtual fields, `month` and `year`, are supported if they don't already exist as real fields but a `date` field does exist",
+    )
+    parser.add_argument(
+        "--priority",
+        type=str,
+        help="file with list priority scores for sequences (strain\tpriority)",
+    )
+    parser.add_argument(
+        "--subsample-seed",
+        help="random number generator seed to allow reproducible sub-sampling (with same input data). Can be number or string.",
+    )
+
+
+def init(context):
+    return Subsample(
+        priorities_fname=context.args.priority,
+        group_by=context.args.group_by,
+        sequences_per_group=context.args.sequences_per_group,
+        seed=context.args.subsample_seed,
+    )
+
+
+class Subsample(BaseFilter):
+    def __init__(self, *, priorities_fname, group_by, sequences_per_group, seed=None):
+        self.priorities_fname = priorities_fname
+        self.discriminators = group_by
+        self.sequences_per_group = sequences_per_group
+        self.seed = seed
+
+    def affected_sequences(self, sequences):
```

Comment:
This feels like the one filter where we really do need all sequences - I'm wondering if subsampling should be a separate action before filtering.